File: src/main/java/com/trustline/trustline/config/security/PrincipalUser.java
package com.trustline.trustline.config.security;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.trustline.trustline.appuser.model.AuthProvider;
import com.trustline.trustline.appuser.model.Role;
import com.trustline.trustline.appuser.model.User;
import lombok.Getter;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import java.util.*;
@Getter
public class PrincipalUser implements UserDetails {
    private final Set<GrantedAuthority> authorities = new HashSet<>();
    private final UUID id;
    private final AuthProvider authProvider;
    private final String username;
    private final boolean enabled;
    private final String password;


    public PrincipalUser(final User user) {
        id = user.getId();
        authProvider = user.getAuthProvider();
        username = user.getEmail();
        enabled = !user.isDeleted() && user.isAccountVerified();
        password = user.getPassword();
        setAuthorities(user.getRoles());
    }

    void setAuthorities(Set<Role> roles) {
        if (roles == null) {
            return;
        }

        for (Role role : roles) {
            authorities.add(new SimpleGrantedAuthority(role.getName()));
            role.getPermissions().stream()
                    .map(p -> new SimpleGrantedAuthority(p.getName()))
                    .forEach(authorities::add);
        }
    }
    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        return authorities;
    }

    @Override
    public String getPassword() {
        return "";
    }

    @Override
    public String getUsername() {
        return username;
    }


    @Override
    public boolean isAccountNonExpired() {
        return true;
    }

    @Override
    public boolean isAccountNonLocked() {
        return true;
    }

    @Override
    public boolean isCredentialsNonExpired() {
        return true;
    }

    @Override
    public boolean isEnabled() {
        return enabled;
    }
}

File: src/main/kotlin/trustline/config/security/PrincipalUser.kt
package trustline.config.security;

import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import trustline.appuser.dto.AuthProvider
import trustline.appuser.model.Role
import trustline.appuser.model.UserModel
import java.util.*

class PrincipalUser(user: UserModel) : UserDetails {

    private val authorities: MutableSet<GrantedAuthority> = HashSet()

    val id: UUID = user.id!!
    val authProvider: AuthProvider = user.authProvider
    private val username: String = user.email
    private val enabled: Boolean = !user.isDeleted && user.isAccountVerified
    private val password: String = user.password!!

    init {
        setAuthorities(user.roles)
    }

    private fun setAuthorities(roles: Set<Role>?) {
        roles ?: return

        for (role in roles) {
            authorities.add(SimpleGrantedAuthority(role.name))

            role.permissions
                .map { SimpleGrantedAuthority(it.name) }
                .forEach(authorities::add)
        }
    }

    override fun getAuthorities(): Collection<GrantedAuthority> = authorities

    override fun getPassword(): String = password

    override fun getUsername(): String = username

    override fun isAccountNonExpired(): Boolean = true

    override fun isAccountNonLocked(): Boolean = true

    override fun isCredentialsNonExpired(): Boolean = true

    override fun isEnabled(): Boolean = enabled
}
